rclcpp/src/rclcpp/experimental/timers_manager.cpp

```diff
@@ -242,17 +244,34 @@ void TimersManager::run_timers()
     // Lock mutex
     std::unique_lock<std::mutex> lock(timers_mutex_);
 
-    std::chrono::nanoseconds time_to_sleep = get_head_timeout_unsafe();
+    std::optional<std::chrono::nanoseconds> time_to_sleep = get_head_timeout_unsafe();
+
+    // If head timer was cancelled, try to reheap and get a new head.
+    // This avoids an edge condition where head timer is cancelled, but other
+    // valid timers remain in the heap.
+    if (!time_to_sleep.has_value()) {
+      // Re-heap to (possibly) move cancelled timer from head of heap. If
+      // entire heap is cancelled, this will still result in a nullopt.
+      TimersHeap locked_heap = weak_timers_heap_.validate_and_lock();
+      locked_heap.heapify();
```

Comment:
The queue is sorted according to the `return a->time_until_trigger() > b->time_until_trigger();` comparison.
What does `time_until_trigger` returns for a cancelled timer?

Reply:
Upon a cancelled timer, `time_until_trigger` returns `std::chrono::nanoseconds::max()`, so resorting should force cancelled timers to be at the bottom of the heap. In a previous iteration of the PR, we had updated the comparison function for the heap to explicitly check for canceled timers and ensure they were on the bottom of the heap, but that was removed since it's obsolete if `std::chrono::nanoseconds::max()` is always returned for a cancelled timer.